File: redis/golang/utils/redis_util.go
package utils

import (
	"context"
	"log"
	"time"

	"github.com/redis/go-redis/v9"
)

type RedisUtil interface {
	GetClient() *redis.Client
	Close(host string, port string)
	Set(ctx context.Context, key string, value interface{}, expiration time.Duration) (string, error)
	Get(ctx context.Context, key string) (string, error)
	Del(ctx context.Context, key string) (int64, error)
}

type redisUtil struct {
	Client *redis.Client
}

func NewRedisUtil(host string, port string, database int) RedisUtil {
	println(time.Now().String()+" redis: connecting to ", host+":"+port)
	rdb := redis.NewClient(&redis.Options{
		Addr:     host + ":" + port,
		Password: "",       // no password set
		DB:       database, // use default DB
	})
	println(time.Now().String()+" redis: connected to", host+":"+port)

	println(time.Now().String()+" redis: pinging to ", host+":"+port)
	ctx := context.Background()
	_, err := rdb.Ping(ctx).Result()
	if err != nil {
		log.Fatalln("redis: error when pinging to:", err)
	}
	println(time.Now().String()+" redis: pinged to ", host+":"+port)

	return &redisUtil{
		Client: rdb,
	}
}

func (util *redisUtil) GetClient() *redis.Client {
	return util.Client
}

func (util *redisUtil) Close(host string, port string) {
	println(time.Now().String()+" redis: closing to ", host+":"+port)
	err := util.Client.Close()
	if err != nil {
		log.Fatalln("redis close connection error:", err)
	}
	println(time.Now().String()+" redis: closed to ", host+":"+port)
}

func (util *redisUtil) Set(ctx context.Context, key string, value interface{}, expiration time.Duration) (string, error) {
	return util.Client.Set(ctx, key, value, expiration).Result()
}

func (util *redisUtil) Get(ctx context.Context, key string) (string, error) {
	return util.Client.Get(ctx, key).Result()
}

func (util *redisUtil) Del(ctx context.Context, key string) (int64, error) {
	return util.Client.Del(ctx, key).Result()
}

File: redis/rust/src/utils/redis_util.rs
use std::process;
use chrono::Local;
use redis::{AsyncCommands, Client, RedisResult};
use redis::aio::MultiplexedConnection;

pub trait RedisUtil {
    async fn set(&mut self, key: &str, value: &str) -> RedisResult<()>;
    async fn get(&mut self, key: &str) -> RedisResult<Option<String>>;
    async fn del(&mut self, key: &str) -> RedisResult<()>;
}

#[derive(Debug)]
pub struct RedisUtilImpl {
    connection: MultiplexedConnection,
}

impl RedisUtilImpl {
    pub async fn new() -> RedisUtilImpl {
        println!("{} redis: connecting to {} {}", Local::now().format("%Y-%m-%d %H:%M:%S%.3f").to_string(), "localhost", "6379");
        let client = match Client::open("redis://localhost:6380/") {
            Ok(client) => client,
            Err(err) => {
                println!("err: {}", err);
                process::exit(1);
            }
        };
        let mut connection = match client.get_multiplexed_async_connection().await {
            Ok(connection) => connection,
            Err(err) => {
                println!("err: {}", err);
                process::exit(1);
            }
        };
        println!("{} redis: connected to {} {}", Local::now().format("%Y-%m-%d %H:%M:%S%.3f").to_string(), "localhost", "6379");

        println!("{} redis: pinging to {} {}", Local::now().format("%Y-%m-%d %H:%M:%S%.3f").to_string(), "localhost", "6379");
        match redis::cmd("PING").query_async::<String>(&mut connection).await {
            Ok(pong) => {
                println!("pong: {}", pong)
            },
            Err(err) => {
                println!("err: {}", err);
                process::exit(1);
            }
        }
        println!("{} redis: pinged to {} {}", Local::now().format("%Y-%m-%d %H:%M:%S%.3f").to_string(), "localhost", "6379");

        RedisUtilImpl {
            connection
        }
    }
}

impl RedisUtil for RedisUtilImpl {
    async fn set(&mut self, key: &str, value: &str) -> RedisResult<()> {
        let _: () = self.connection.set(key, value).await?;
        Ok(())
    }

    async fn get(&mut self, key: &str) -> RedisResult<Option<String>> {
        let result = self.connection.get(key).await?;
        Ok(result)
    }

    async fn del(&mut self, key: &str) -> RedisResult<()> {
        let _: () = self.connection.del(key).await?;
        Ok(())
    }
}